blocks/search.py

```diff
@@ -0,0 +1,304 @@
+"""The beam search module."""
+from collections import OrderedDict
+from six.moves import range
+
+import numpy
+
+from theano import config, function, tensor
+
+from blocks.bricks.sequence_generators import SequenceGenerator
+from blocks.filter import VariableFilter, get_application_call, get_brick
+from blocks.graph import ComputationGraph
+from blocks.roles import INPUT, OUTPUT
+
+floatX = config.floatX
+
+
+class BeamSearch(object):
+    """Approximate search for the most likely sequence.
+
+    Beam search is an approximate algorithm for finding :math:`y^* =
+    argmax_y P(y|c)`, where :math:`y` is an output sequence, :math:`c` are
+    the contexts, :math:`P` is the output distribution of a
+    :class:`.SequenceGenerator`. At each step it considers :math:`k`
+    candidate sequence prefixes. :math:`k` is called the beam size, and the
+    sequence are called the beam. The sequences are replaced with their
+    :math:`k` most probable continuations, and this is repeated until
+    end-of-line symbol is met.
+
+    Parameters
+    ----------
+    beam_size : int
+        The beam size.
+    samples : :class:`~theano.Variable`
+        An output of a sampling computation graph built by
+        :meth:`~blocks.brick.SequenceGenerator.generate`, the one
+        corresponding to sampled sequences.
+
+    See Also
+    --------
+    :class:`.SequenceGenerator`
+
+    Notes
+    -----
+    Sequence generator should use an emitter which has `probs` method
+    e.g. :class:`SoftmaxEmitter`.
+
+    """
+    def __init__(self, beam_size, samples):
+        self.beam_size = beam_size
+
+        # Extracting information from the sampling computation graph
+        cg = ComputationGraph(samples)
+        self.inputs = cg.inputs
+        self.generator = get_brick(samples)
+        if not isinstance(self.generator, SequenceGenerator):
+            raise ValueError
+        self.generate_call = get_application_call(samples)
+        if (not self.generate_call.application ==
+                self.generator.generate):
+            raise ValueError
+        self.inner_cg = ComputationGraph(self.generate_call.inner_outputs)
+
+        # Fetching names from the sequence generator
+        self.context_names = self.generator.generate.contexts
+        self.state_names = self.generator.generate.states
+
+        # Parsing the inner computation graph of sampling scan
+        self.contexts = [
+            VariableFilter(bricks=[self.generator], name='^' + name + '$',
+                           roles=[INPUT])(self.inner_cg)[0]
+            for name in self.context_names]
```

Comment:
Are are right, but I will leave it for a refactoring that should follow after #391.